runtime/argument_validation.go

```diff
@@ -0,0 +1,354 @@
+//
+func valueConformsToDynamicType(value interpreter.Value, dynamicType interpreter.DynamicType) (ok bool) {
+
+	// Currently the type-vs-value mismatch could only happen for composite types, as
+	// that's the only place where the type is loaded from the runtime, based on the name,
+	// rather than based on the value structure of the user input.
+	// Hence the possible set of malformed values are:
+	//     - Composite types.
+	//     - Arrays/Dictionaries/Optionals of composite types.
+
+	switch typ := dynamicType.(type) {
+	case interpreter.ArrayDynamicType:
+		ok = valueConformsToArrayDynamicType(value, typ)
+	case interpreter.CompositeDynamicType:
+		ok = valueConformsToSemaType(value, typ.StaticType)
+	case interpreter.DictionaryDynamicType:
+		ok = valueConformsToDictionaryDynamicType(value, typ)
+	case interpreter.SomeDynamicType:
+		ok = valueConformsToSomeDynamicType(value, typ)
+
+	// Following types are guaranteed to be decoded correctly by the json.decode().
+	// However, this additional layer is added to protect the Cadence runtime,
+	// since the decoding currently relies on the host env, and it is external to Cadence.
+	case interpreter.StringDynamicType:
+		_, ok = value.(*interpreter.StringValue)
+	case interpreter.BoolDynamicType:
+		_, ok = value.(interpreter.BoolValue)
+	case interpreter.NumberDynamicType:
+		ok = valueConformsToSemaType(value, typ.StaticType)
+	case interpreter.NilDynamicType:
+		_, ok = value.(interpreter.NilValue)
+	case interpreter.AddressDynamicType:
+		_, ok = value.(interpreter.AddressValue)
+	case interpreter.MetaTypeDynamicType:
+		_, ok = value.(interpreter.TypeValue)
+	case interpreter.CapabilityDynamicType:
+		_, ok = value.(interpreter.CapabilityValue)
+	case interpreter.PrivatePathDynamicType, interpreter.PublicPathDynamicType, interpreter.StoragePathDynamicType:
+		_, ok = value.(interpreter.PathValue)
+
+	// Following types cannot be used as arguments to a script/transaction.
+	// However, still validate and allow wherever possible, so that this validation
+	// is less conservative and would seamlessly cater future changes.
+	case interpreter.FunctionDynamicType:
+		_, ok = value.(interpreter.FunctionValue)
+	case interpreter.BlockDynamicType:
+		_, ok = value.(interpreter.BlockValue)
+	case interpreter.DeployedContractDynamicType:
+		_, ok = value.(interpreter.DeployedContractValue)
+	case interpreter.StorageReferenceDynamicType:
+		// Currently this only checks whether the the value is `interpreter.StorageReferenceValue`.
+		// It doesn't check whether the 'referenced' value conforms to the innerType.
+		// TODO: add support for checking the referenced value conformance (if importing is supported).
+		_, ok = value.(*interpreter.StorageReferenceValue)
+	case interpreter.EphemeralReferenceDynamicType:
+		ok = valueConformsToEphemeralReferenceDynamicType(value, typ)
+	case interpreter.VoidDynamicType:
+		// Void type cannot have a value.
+		ok = false
```

Comment:
There actually is a value, `VoidValue`. It's a bit counter-intuitive, but it is e.g returned from a function that returns `Void`, and could be passed to a function with a `Void` parameter type